gnucash/register/register-gnome/completioncell-gnome.c

```diff
@@ -0,0 +1,1019 @@
+        }
+        box->occurrence++;
+    }
+}
+
+void
+gnc_completion_cell_set_value (CompletionCell* cell, const char* str)
+{
+    if (!cell || !str)
+
+    gnc_basic_cell_set_value (&cell->cell, str);
+}
+
+static inline void
+list_store_append (GtkListStore *store, char* string,
+                   char* markup, gint weight)
+{
+    GtkTreeIter iter;
+
+    g_return_if_fail (store);
+    g_return_if_fail (string);
+    g_return_if_fail (markup);
+
+    gtk_list_store_append (store, &iter);
+
+    gtk_list_store_set (store, &iter, TEXT_COL, string,
+                                      TEXT_MARKUP_COL, markup,
+                                      WEIGHT_COL, weight, -1);
+}
+
+static char*
+normalize_and_fold (char* utf8_string)
+{
+    char *normalized = g_utf8_normalize (utf8_string, -1, G_NORMALIZE_NFC);
+    if (!normalized)
+        return NULL;
+
+    char *folded = g_utf8_casefold (normalized, -1);
+    g_free (normalized);
+    return folded;
+}
+
+static void
+add_item (gpointer key, gpointer value, gpointer user_data)
+{
+    PopBox* box = user_data;
+    gchar *hash_entry = g_strdup (key);
+
+    if (hash_entry && *hash_entry)
+    {
+        gnc_utf8_strip_invalid_and_controls (hash_entry);
+        gchar *key_norm_fold = normalize_and_fold (hash_entry);
+        gchar *found = g_strstr_len (key_norm_fold, -1, box->newval);
+
+        if (found)
+        {
+            gchar *markup = NULL, *prefix = NULL, *match = NULL, *suffix = NULL;
+            gint prefix_len;
+            gint occurrence_difference;
+            gint weight;
+            gboolean have_boundary = FALSE;
+            glong len = g_utf8_strlen (box->newval, -1);
+            gulong loc = g_utf8_pointer_to_offset (key_norm_fold, found);
+            OccurrenceData *od = (OccurrenceData*)value;
+
+            if (box->register_is_reversed)
+                occurrence_difference = od->occurrence + 1;
+            else
+                occurrence_difference = box->occurrence - od->occurrence;
+
+            if (loc > 0)
+                prefix = g_utf8_substring (hash_entry, 0, loc);
+            else
+                prefix = g_strdup ("");
+            prefix_len = g_utf8_strlen (prefix, -1);
+
+            match = g_utf8_substring (hash_entry, loc, loc + len);
+
+            if (loc > 1)
+            {
+                gunichar prev = g_utf8_get_char (g_utf8_offset_to_pointer (hash_entry, loc - 1));
+                if (prev && (g_unichar_isspace (prev) || g_unichar_ispunct (prev)))
+                    have_boundary = TRUE;
+            }
+
+            suffix = g_utf8_substring (hash_entry, loc + len, g_utf8_strlen (hash_entry, -1));
+
+            if (g_utf8_strlen (suffix, -1) == 0)
+                markup = g_markup_printf_escaped ("%s<b>%s</b>%s", prefix, match, " ");
+            else
+                markup = g_markup_printf_escaped ("%s<b>%s</b>%s%s", prefix, match, suffix, " ");
```

Comment:
If I understand the code behind `g_utf8_substring` correctly it will always return a string. If there's no suffix it will be the empty string `""`. That's a valid value for the `%s` positional parameter, no ? Any reason to make different format strings based on this other than that ?

Reply:
Also, why do you end the markup string with a space character ?

Reply:
> If I understand the code behind `g_utf8_substring` correctly it will always return a string. If there's no suffix it will be the empty string `""`. That's a valid value for the `%s` positional parameter, no ? Any reason to make different format strings based on this other than that ?

Yes this can be reduced to the one line, missed testing that when I added the space.

Reply:
> Also, why do you end the markup string with a space character ?

For some reason, could not see why and still cant, without the space when you type in the exact match so you end up with 'Don't autocomplete' and say 'best long text', the popup increases in size with a blank area below the second entry.

Every thing I looked at seemed correct, I will have another look.